### DISCUSSION TITLE
Can Ctrlr handle Windows bitmaps as image type?

### DISCUSSION THREAD

**Discussion by shooking** [RomanKubiak/ctrlr#304] [Q&A]
I dont think it can but not sure how to be sure.

Take the following:
```
pi@raspberrypi:~/Software/ZoomPedal/ZD2 $ file BLANK.BMP 
BLANK.BMP: PC bitmap, Windows 3.x format, 47 x 30 x 1
pi@raspberrypi:~/Software/ZoomPedal/ZD2 $ file BLANK1.BMP 
BLANK1.BMP: PNG image data, 47 x 30, 1-bit grayscale, non-interlaced
```
So if I load the BLANK.BMP as a resource then Ctrlr suggests it is data
If I convert this to PNG, Ctrlr loads it.
Of course if I change its name from PNG to BMP the "file" type isnt fooled and Ctrlr just loads it.

Hence I conclude Ctrlr cannot open BMPs? At least not Windows ones.

Anyone managed to load them?
Sure I can easily convert them from BMP to PNG - except the pedal I am using stores Windows BMs and ideally I would love to extract and load the BMPs directly into Ctrlr.

Best regards
Steve

**Comment by shooking** [comment]
One more thing I notice - even if I use a PNG I hit some assertion error when I click on panel (and redraw happens)

```
JUCE Assertion failure in juce_GraphicsContext.cpp:41

Thread 1 "Ctrlr" received signal SIGTRAP, Trace/breakpoint trap.
0xb677520c in kill () from /lib/arm-linux-gnueabihf/libc.so.6
(gdb) where
#0  0xb677520c in kill () from /lib/arm-linux-gnueabihf/libc.so.6
#1  0x0049a268 in juce::(anonymous namespace)::coordsToRectangle<int> (x=0, y=0, w=96, h=-14)
    at ../../JUCE/modules/juce_graphics/contexts/juce_GraphicsContext.cpp:38

...
```

so something is miscalculating the image bounds and debugger catches this - if I didnt run Ctrlr under a debugger it would likely keep crashing. As it stands I at least get some idea why it crashes - but cannot work out why getting -ve bounds.

**Reply by shooking** [reply]
OK I worked that 2nd Gotcha out.
By default the constructor makes the image 96, 32 but this doesnt seem large enough - leading to assertion during construction - if I didnt have a debugger it would crash the panel and I would never be able to load it.
Because I catch the exception I can eventually trace it back.
Now - the fix I found so far then is to pray you have the button selected (so that the component properties are visible).
DONT click on the paint canvas - scroll and make the rectangle larger in 280 94 96 96
(it was previously 96 32 as per constructor).
Then the panel doesnt crash!!

```
CtrlrImageButton::CtrlrImageButton (CtrlrModulator &owner)
    : CtrlrComponent(owner),
      ctrlrButton (0)
{
	valueMap = new CtrlrValueMap();
    addAndMakeVisible (ctrlrButton = new CtrlrCustomButtonInternal (*this));

    //[UserPreSize]
	setBufferedToImage (true);
	ctrlrButton->addListener (this);
	setProperty (Ids::uiImageButtonMode, 0);
	setProperty (Ids::uiImageButtonResource, "");
	setProperty (Ids::resourceImageWidth, 32);
	setProperty (Ids::resourceImageHeight, 32);
	setProperty (Ids::resourceImagePaintMode, 36);
	setProperty (Ids::resourceImageOrientation, 1);
	setProperty (Ids::uiImageButtonTextColour, "0xff000000");
	setProperty (Ids::uiImageButtonContent, "");
	setProperty (Ids::uiImageButtonTextPosition, 4);
	setProperty (Ids::uiImageButtonTextWidth, 0);
	setProperty (Ids::uiImageButtonTextHeight, 32);
	setProperty (Ids::uiButtonTextFont, Font(12).toString());
	setProperty (Ids::uiButtonTextJustification, "centred");
	setProperty (Ids::uiImageButtonTextPosition, "bottom");
	setProperty (Ids::uiImageButtonContent, "False\nTrue");
	setProperty (Ids::uiButtonRepeat, false);
	setProperty (Ids::uiButtonRepeatRate, 100);
    //[/UserPreSize]

    setSize (96, 32);

    //[Constructor] You can add your own custom stuff here..
    //[/Constructor]
}
```
I just checked in git - code is same. So maybe others might be hitting issues if they try to create a new UIButton?

**Comment by dnaldoog** [comment]
I tried doing that, but I don't think one can. It registers BMP as Data, but PNG as Image.

But what I did discover, which is rather cool, is that you can load text files into a _uiCustomComponent_.

<pre>
    local t = resources:getResource("mytextfile") -- mytextfile.txt
    local x = t:asText()
    g:drawText(x, comp:getWidth() / 3, comp:getHeight() / 3, 200, 50, Justification(Justification.left), true)
</pre>

Trying to think how this could be useful. 🐱‍🐉

[Load Text as a Resource_1_0_2021-05-21_09-07.zip](https://github.com/RomanKubiak/ctrlr/files/6519259/Load.Text.as.a.Resource_1_0_2021-05-21_09-07.zip)

**Reply by dnaldoog** [reply]
You can in fact load [BMP ](http://www.fastgraph.com/help/bmp_header_format.html) as Data!

<pre>
local x = resources:getResource("Untitled"):asData() -- Untitled.bmp
local m=MemoryBlock(x)

local bmpHeader=54 -- bytes
local imgByte=m:getByte(34)
local bwidth=m:getByte(18)
local bheight=m:getByte(22)
local counter=0
bitmap_header=m:getRange(0,bmpHeader):toHexString(1)
imageData=m:getRange(bmpHeader,m:getSize()-bmpHeader)):toHexString(1)

--[[
offset size description 54 byte header
0 2 signature, must be 4D42 hex
2 4 size of BMP file in bytes (unreliable)
6 2 reserved, must be zero
8 2 reserved, must be zero
10 4 offset to start of image data in bytes
14 4 size of BITMAPINFOHEADER structure, must be 40
18 4 image width in pixels
22 4 image height in pixels
26 2 number of planes in the image, must be 1
28 2 number of bits per pixel (1, 4, 8, or 24)
30 4 compression type (0=none, 1=RLE-8, 2=RLE-4)
34 4 size of image data in bytes (including padding)
38 4 horizontal resolution in pixels per meter (unreliable)
42 4 vertical resolution in pixels per meter (unreliable)
46 4 number of colors in image, or zero
50 4 number of important colors, or zero
--]]
</pre>

**Reply by shooking** [reply]
Nice to see some Lua (I was also seeing there is a struct I can use).
I extract the BM direct from the Sysex (but currently via Python or C++) and write the BM.
I can use "covert" externally to turn it into a PNG - would prefer the handle the BM direct - or in a way that is processable via Lua to a format that Ctrlr can understand.

Is there a way to take that imageData above and draw it on the Canvas -- ideally I want to create a button with this BM in it.
Click, turn off (and change some colour or something)
Click, turn it on (and toggle colour).

I also want to do it dynamically.
To be fair even tkInter requires me to convert the BM into a PNG.
I can see that BM was quite easy to be used as a hacking vector. Maybe that's why it is not natively supported.

**Reply by dnaldoog** [reply]
[I wrote a program that emulates XPM](https://ctrlr.org/forums/topic/creating-images-using-lua/#post-120273) some time back on the old forums.

If you knew the byte order and therefore how to unpack the bytes of a bitmap now that the binary data is attainable in Ctrlr, I guess you could then loop through using something like:

<pre>
for i = 0, image_height_in_pixels do
    for j = 0, image_width_in_ pixels do
        if m:getByte(j) == 0xff then -- m is MemoryBlock of the BMP
            g:setColour(white)
            g:fillRect(j, i, 2, 2)
        else
            g:setColour(black)
            g:fillRect(j, i, 2, 2)
        end
    end
end
</pre>

to recreate a vector image in _uiComponent_. That would be pretty wild. Pseudo code above of course!

**Reply by shooking** [reply]
neat - the old forum banned me - so I cannot open the panel you posted. Let me try to hack the above. Thanks again.

**Reply by dnaldoog** [reply]
Banned! ??? I'm not sure I should be posting this here for your perusal in that case! 😒

[LCD xpm Matrix_1_0_0.zip](https://github.com/RomanKubiak/ctrlr/files/6525355/LCD.xpm.Matrix_1_0_0.zip)
